tutorials/cloud-run-with-pomerium-for-end-user-access/index.md

```diff
@@ -0,0 +1,190 @@
+description: Learn how to deploy Pomerium to Cloud Run and use it to protect other endpoints with authorization headers.
+author: travisgroth,desimone
+tags: Cloud Run, Pomerium
+date_published: 2020-08-25
+---
+
+This guide covers how to deploy Pomerium to Cloud Run, providing end-user authentication and authorization to other endpoints. The resulting configuration
+permits users with `@gmail.com` addresses to access an instance of `httpbin.org` hosted on Cloud Run.
+
+[Pomerium](https://www.pomerium.com) is an open source identity-aware proxy that enables secure access to internal applications. Pomerium provides a standardized
+interface to add access control to applications regardless of whether the application itself has authorization or authentication built in. Pomerium provides a 
+gateway for requests and can be used in situations where you'd typically consider using a VPN.
+
+Unlike [Cloud IAP](https://cloud.google.com/iap), Pomerium supports non-Google identity providers. You can also run Pomerium outside Google Cloud (such as on 
+other cloud providers and on-premises), and still use it to route or authorize traffic to Google Cloud targets such as Cloud Run or Cloud Functions.
+
+This guide assumes that you have editor access to a Google Cloud project that can be used for isolated testing and a DNS zone that you are able to control.
+DNS does not need to be inside Google Cloud for the example to work.
+
+## How it works
+
+Services on [Cloud Run](https://cloud.google.com/run) and [Cloud Functions](https://cloud.google.com/functions) can be restricted to only permit access with a 
+properly signed [identity token](https://cloud.google.com/run/docs/authenticating/service-to-service). This allows requests from other services running on
+Google Cloud or elsewhere to be securely authorized despite the endpoints being public.
+
+Browsers are not able to add these identity tokens to the `Authorization` header, so a proxy is required to authenticate end users to your Cloud Run service.
+Pomerium can perform this role by generating compatible tokens on behalf of end users and then acting as a proxy for their requests.
+
+You perform the following steps:
+
+- Add an IAM policy delegating `roles/run.invoker` permissions to a service account.
+- Run Pomerium with access to a key for the corresponding service account.
+- Publish DNS records for each protected application pointing to Pomerium.
+- Configure Pomerium with appropriate policy and `enable_google_cloud_serverless_authentication`.
+
+The protected application delegates trust to a Google Cloud service account, which Pomerium runs as, and Pomerium performs user-based authorization for each
+route. This turns Pomerium into a bridge between a user-centric and a service-centric authorization model.
+
+## Setup
+
+To deploy Pomerium to Cloud Run, a [special image](https://console.cloud.google.com/gcr/images/pomerium-io/GLOBAL/pomerium) is available at
```

Comment:
@travisgroth is this still the case? maybe we should change this as soon as Cloud Run integration is available in the main image?

Reply:
The bootstrap to provide secrets from Secret Manager is not going to be in the main image for the time being.

Reply:
Almost forgot - we have to publish an image for Cloud Run, as our regular image is not hosted in gcr.  Cloud Run managed seems to only allows deployments from gcr.

Reply:
Ack. For the purposes of the tutorial it's not a bad idea to circumvent the secrets manager entirely, and store the config.yaml in env vars as well.

Reply:
Since environment variables in Cloud Run are considered insecure, I'm hesitant to encourage storing IdP credentials and other sensitive values that way in broadly consumed documentation.  We added the Secret Manager wrapper specifically to avoid setting a bad example.

The builds are the same and maintained with our CI, so this isn't a second class image.  The wrapper isn't worth putting into the normal image since most of our other container users have additional options for pulling in secrets.  In Cloud Run we recommend using the Secrets Manager wrapper until there is an alternative.

Reply:
ack. BTW you can recommend usage of https://github.com/GoogleCloudPlatform/berglas which is a supported tool that wraps a process and has its on Secret Manager reference syntax. It retrieves secrets and sets them as env on runtime.